runtime/fundamentals/ffi.md

```diff
@@ -0,0 +1,256 @@
+code performance and capabilities directly into your Deno applications.
+
+## Introduction to FFI
+
+FFI provides a bridge between Deno's JavaScript runtime and native code. This
+allows you to:
+
+- Use existing native libraries within your Deno applications
+- Implement performance-critical code in languages like Rust or C
+- Access operating system APIs and hardware features not directly available in
+  JavaScript
+
+Deno's FFI implementation is based on the `Deno.dlopen` API, which loads dynamic
+libraries and creates JavaScript bindings to the functions they export.
+
+## Security considerations
+
+FFI requires explicit permission using the
+[`--allow-ffi`](/runtime/fundamentals/security#ffi-foreign-function-interface)
+flag, as native code runs outside of Deno's security sandbox:
+
+```sh
+deno run --allow-ffi my_ffi_script.ts
+```
+
+:::info
+
+⚠️ **Important security warning**: Unlike JavaScript code running in the Deno
+sandbox, native libraries loaded via FFI have the same access level as the Deno
+process itself. This means they can:
+
+- Access the filesystem
+- Make network connections
+- Access environment variables
+- Execute system commands
+
+Always ensure you trust the native libraries you're loading through FFI.
+
+:::
+
+## Basic usage
+
+The basic pattern for using FFI in Deno involves:
+
+1. Defining the interface for the native functions you want to call
+2. Loading the dynamic library using `Deno.dlopen()`
+3. Calling the loaded functions
+
+Here's a simple example loading a C library:
+
+```ts
+// Define the symbols (functions) we want to use from the library
+const libName = {
+  windows: "example.dll",
+  linux: "libexample.so",
+  darwin: "libexample.dylib",
+}[Deno.build.os];
+
+// Define the function signatures
+const signatures = {
+  add: { parameters: ["i32", "i32"], result: "i32" },
+} as const;
+
+// Open the library
+const dylib = Deno.dlopen(libName, signatures);
+
+// Use the function
+const result = dylib.symbols.add(5, 3);
+console.log(result); // 8
+
+// Close the library when done
+dylib.close();
+```
+
+## Supported types
+
+Deno's FFI supports a variety of data types for parameters and return values:
+
+| Type                                                               | Description                              |
+| ------------------------------------------------------------------ | ---------------------------------------- |
+| `i8`, `u8`, `i16`, `u16`, `i32`, `u32`, `i64`, `u64`, `f32`, `f64` | Integer and float types of various sizes |
+| `pointer`                                                          | Raw memory pointer                       |
+| `function`                                                         | Function pointer                         |
+| `buffer`                                                           | Pointer to a buffer (typed array)        |
+| `struct`                                                           | C struct representation                  |
```

Comment:
Maybe yoink this table from the manual? It compares each type against a compiled lang equivalent.

https://github.com/denoland/docs/blob/df998002fa76f3e6a788d600a6886c97461bca71/runtime/reference/deno_namespace_apis.mdx?plain=1#L669-L688